// node-smushit/lib/smushit.js
var fs = require("fs"),
    url = require("url"),
	http = require("http");


exports.SMUSH_SERVICE = 'http://www.smushit.com/ysmush.it/ws.php';
	
	//构建上传文件的请求body
var	buildRequestBody = function(fullName, uploadIdentifier, params){
		var boundary = '------multipartformboundary' + (new Date).getTime();
		var dashdash = '--';
		var crlf     = '\r\n';
	 
		/* Build RFC2388. */
		var builder = '';


		builder += dashdash;
		builder += boundary;
		builder += crlf;
	 
		builder += 'Content-Disposition: form-data; name="'+ uploadIdentifier +'"';
		//支持文件名为中文
		builder += '; filename="' + encodeURIComponent(fullName.replace(/.+\//, '')) + '"';
		builder += crlf;
	 
		builder += 'Content-Type: application/octet-stream';
		builder += crlf;
		builder += crlf;
	 
		/* 写入文件 */
		builder += fs.readFileSync(fullName, "binary");
		builder += crlf;
	 
		/* 传递额外参数 */
		for(var i in params){
			if(params.hasOwnProperty(i)){
				builder += dashdash;
				builder += boundary;
				builder += crlf;
		 
				builder += 'Content-Disposition: form-data; name="'+ i +'"';
				builder += crlf;
				builder += crlf;
				//支持参数为中文
				builder += encodeURIComponent(params[i]);
				builder += crlf;
			}
		}
		
	 
		/* 写入边界 */
		builder += dashdash;
		builder += boundary;
		builder += dashdash;
		builder += crlf;
		//console.log(builder);
		return {
			contentType: 'multipart/form-data; boundary=' + boundary,
			builder: builder
		}
}

var smushit = function(str, success, fail, customService){
	var seriveEndpoint = customService || exports.SMUSH_SERVICE,
	
	serviceParsed = url.parse(seriveEndpoint),
	options = {
	  host: serviceParsed.hostname,
	  port: serviceParsed.port,
	  path: serviceParsed.pathname
	},
	httpRequest,
	onRequestCompleted = function(response) {
	  var resBody = '';
	  response.on('data', function(chunk) {
		resBody += chunk;
	  });
	  response.on('end', function() {
		  success && success(resBody);
	  });
	};

	if(str.match(/https?:\/\//)){
		options.method = "GET";
		options.path += "?img=" + encodeURIComponent(str);
		httpRequest = http.get(options, onRequestCompleted, onR);
	}else{
		var postData = buildRequestBody(str, "files");
		options.method = "POST";
		options.headers =  {
		  'Content-Type': postData.contentType,
		  'Content-Length': postData.builder.length
	    }
		httpRequest = http.request(options, onRequestCompleted);
		httpRequest.write(postData.builder, "binary");
		httpRequest.end();
	}
	
	if(fail){
		httpRequest.on('error', fail);
	}
};

exports.smushit = smushit;
